src/__rt/mod.rs

```diff
@@ -0,0 +1,56 @@
+mod future;
+mod stream;
+mod pinned_future;
+mod pinned_stream;
+
+use std::cell::Cell;
+use std::ptr;
+use futures::task;
+
+pub use self::future::*;
+pub use self::stream::*;
+pub use self::pinned_future::*;
+pub use self::pinned_stream::*;
+
+pub use futures::prelude::{Async, Future, Stream};
+pub use stable::{StableFuture, StableStream};
+
+pub extern crate std;
+pub extern crate anchor_experiment;
+
+pub use std::ops::Generator;
+
+#[rustc_on_unimplemented = "async functions must return a `Result` or \
+                            a typedef of `Result`"]
+pub trait IsResult {
+    type Ok;
+    type Err;
+
+    fn into_result(self) -> Result<Self::Ok, Self::Err>;
+}
+impl<T, E> IsResult for Result<T, E> {
+    type Ok = T;
+    type Err = E;
+
+    fn into_result(self) -> Result<Self::Ok, Self::Err> { self }
+}
+
+pub fn diverge<T>() -> T { loop {} }
+
+/// Uninhabited type to allow `await!` to work across both `async` and
+/// `async_stream`.
+pub enum Mu {}
+
+thread_local!(static CTX: Cell<*mut task::Context<'static>> = Cell::new(ptr::null_mut()));
+
+struct Reset<'a>(*mut task::Context<'static>, &'a Cell<*mut task::Context<'static>>);
+
+impl<'a> Drop for Reset<'a> {
+    fn drop(&mut self) {
+        self.1.set(self.0);
+    }
+}
+
+pub fn get_ctx() -> *mut task::Context<'static> {
```

Comment:
In order for the usage of this above to be safe, you have to replace the existing pointer with a nullptr. This prevents you from using `await` in, for example, the impl of a `Future::poll` function which is currently being awaited, causing aliasing. Does that make sense?

Reply:
Once you're done with the pointer, you can put it back.

Reply:
Actually, I think you could use this approach to take a `with_ctx` closure here and then make this whole API safe, removing the need for `unsafe` code in the generated code (which is harder to read/modify/validate than an isolated `unsafe` abstraction.